service/status-report/main.go

```diff
@@ -118,29 +126,63 @@ func main() {
+	if err := updateCSIConfig(ctx, cl, providerClient, storageClient, operatorNamespace); err != nil {
+		klog.Exitf("Failed to update csi config: %v", err)
+	}
+}
+
+func updateCSIConfig(ctx context.Context,
+	cl client.Client,
+	providerClient *providerclient.OCSProviderClient,
+	storageClient *v1alpha1.StorageClient,
+	operatorNamespace string) error {
 	scResponse, err := providerClient.GetStorageConfig(ctx, storageClient.Status.ConsumerID)
 	if err != nil {
-		klog.Exitf("Failed to get StorageConfig of storageClient %v: %v", storageClient.Status.ConsumerID, err)
+		return fmt.Errorf("failed to get StorageConfig of storageClient %v: %v", storageClient.Status.ConsumerID, err)
 	}
 	for _, eResource := range scResponse.ExternalResource {
-		if eResource.Kind == "ConfigMap" && eResource.Name == "rook-ceph-mon-endpoints" {
+		if !utils.DelegateCSI && eResource.Kind == "ConfigMap" && eResource.Name == "rook-ceph-mon-endpoints" {
 			monitorIps, err := csi.ExtractMonitor(eResource.Data)
 			if err != nil {
-				klog.Exitf("Failed to extract monitor data for storageClient %v: %v", storageClient.Status.ConsumerID, err)
+				return fmt.Errorf("failed to extract monitor data for storageClient %v: %v", storageClient.Status.ConsumerID, err)
 			}
+			var csiClusterConfigEntry = new(csi.ClusterConfigEntry)
 			csiClusterConfigEntry.Monitors = append(csiClusterConfigEntry.Monitors, monitorIps...)
+			cc := csi.ClusterConfig{
+				Client:    cl,
+				Namespace: operatorNamespace,
+				Ctx:       ctx,
+			}
+			err = cc.UpdateMonConfigMap("", storageClient.Status.ConsumerID, csiClusterConfigEntry)
+			if err != nil {
+				return fmt.Errorf("failed to update mon configmap for storageClient %v: %v", storageClient.Status.ConsumerID, err)
+			}
+		} else if utils.DelegateCSI && eResource.Kind == "CephConnection" {
+			desiredCephConnectionSpec := &csiopv1a1.CephConnectionSpec{}
+			if err := json.Unmarshal(eResource.Data, &desiredCephConnectionSpec); err != nil {
+				return fmt.Errorf("failed to unmarshall cephConnectionSpec: %v", err)
+			}
+
+			cephConnection := &csiopv1a1.CephConnection{}
+			cephConnection.Name = storageClient.Name
+			cephConnection.Namespace = operatorNamespace
+			if err := cl.Get(ctx, client.ObjectKeyFromObject(cephConnection), cephConnection); err != nil {
+				return fmt.Errorf("failed to get csi cephConnection resource: %v", err)
+			}
+
+			if !reflect.DeepEqual(cephConnection.Spec, desiredCephConnectionSpec) {
+				cephConnectionCopy := &csiopv1a1.CephConnection{}
+				cephConnection.ObjectMeta.DeepCopyInto(&cephConnectionCopy.ObjectMeta)
+				desiredCephConnectionSpec.DeepCopyInto(&cephConnectionCopy.Spec)
+				// patch is being used to ensure spec is accurate at this point of time even if there are reconciles happening which change resourceversion
+				if err := cl.Patch(ctx, cephConnection, client.MergeFrom(cephConnectionCopy)); err != nil {
+					return fmt.Errorf("failed to patch csi cephConnectionSpec: %v", err)
+				}
+			}
```

Comment:
This code should not be part of the reporter. Please move it into the appropriate controller

Reply:
We need to update endpoints received as a response to heartbeat when the hash feature is in then controller will take over.